urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/client/UserImpl.java

```diff
@@ -0,0 +1,500 @@
+import constants.BookSide;
+import exceptions.ExceptionHandler;
+import gui.UserDisplayManager;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import price.Price;
+import publishers.CancelMessage;
+import publishers.FillMessage;
+import tradable.TradableDTO;
+
+/**
+ * This class represents a real user in the trading system.
+ *
+ * @author Kevin Patrick Westropp
+ * @version StockExVirtuoso.Version1.0
+ */
+public class UserImpl implements User {
+
+    /**
+     * String of User name.
+     */
+    private String user;
+    /**
+     * Connection Id provided by system upon connection.
+     */
+    private long connectionId;
+    /**
+     * List of currently available stocks.
+     */
+    private ArrayList<String> stocks = new ArrayList();
+    /**
+     * List containing all orders for this user.
+     */
+    private ArrayList<TradableUserData> allOrders = new ArrayList();
+    /**
+     * Position Object which holds values of users stocks, costs, etc.
+     */
+    private Position userPosition;
+    /**
+     * Reference to the facade for the user interaction.
+     */
+    private UserDisplayManager myManager;
+
+    /**
+     * Public Constructor method.
+     *
+     * @param userName
+     * @throws Exception
+     */
+    public UserImpl(String userName) throws Exception {
+        setUserName(userName);
+        userPosition = new Position();
+    }
+
+    /**
+     * Private set method to set the user name.
+     *
+     * @param userName
+     * @throws Exception
+     */
+    private void setUserName(String userName) throws Exception {
+        if (ExceptionHandler.checkString(userName, "client.UserImpl#setUserName.")) {
+            user = userName;
+        }
+    }
+
+    /**
+     * Public get method to return user name.
+     *
+     * @return user name
+     */
+    @Override
+    public String getUserName() {
+        return user;
+    }
+
+    /**
+     * This method will call the user's display manager and position to update
+     * the last sale.
+     *
+     * @param Product
+     * @param p
+     * @param v
+     */
+    @Override
+    public void acceptLastSale(String Product, Price p, int v) {
+        try {
+            myManager.updateLastSale(Product, p, v);
+            userPosition.updateLastSale(Product, p);
+        } catch (Exception ex) {
+            System.out.println("Caught error: " + ex.getMessage());
+        }
+    }
+
+    /**
+     * This method will display the Fill Message in the market display and will
+     * forward the data to the Position object.
+     *
+     * @param fm fill message
+     */
+    @Override
+    public void acceptMessage(FillMessage fm) {
+        try {
+            Timestamp time = new Timestamp(System.currentTimeMillis());
+            StringBuilder sb = new StringBuilder();
+            sb.append(String.format("{%s} Fill Message: %s %s %s at %s", time.toString(), fm.getSide(), fm.getVolume(), fm.getProduct(), fm.getPrice().toString()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;getVolume&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 112. Consider formatting this data with &#39;%d&#39; instead.